Add Settings Exporter tool (copy settings to INI, etc)

rwf-rr opened the pull request:
Add a Contrib tool to export the current settings (Options) from the registry to the OpenRails.ini file (and vice versa). Depends on the settings refactoring done by [PR 1030](https://github.com/openrails/openrails/pull/1030) (merged).

Example usage, showing help:

```
C:\Users\roger\source\repos\rwf-openrails\Program>Contrib.SettingsExporter /help

Usage: Contrib.SettingsExporter <from> <to>
  <from>     Specify the source to load settings from. It may be:
               INI  : use the default INI file OpenRails.ini.
               REG  : use the default registry key SOFTWARE\OpenRails\ORTS.
               path : a specific INI file, relative to the OpenRails main folder. Must end with ".ini".
               key  : a specific registry key, relative to the HKEY_CURRENT_USER key.
  <to>       Specify the destination to save the settings to. Similar to <from>.
  /h, /help  Show this help.

This utility reads the Settings (Options) from one location, and exports them to another location.
It creates a backup of any settings that will be overwritten. Example:
  <installfolder>\OpenRails.ini.bak
  HKEY_CURRENT_USER\SOFTWARE\OpenRails\ORTS-Backup
This utility is intended to:
- Create an INI file that has the same settings as what is in the registry.
  Example: Contrib.SettingsExporter REG INI
- Copy the settings from an INI file back into the registry, so that other installations
  use the same settings.
  Example: Contrib.SettingsExporter INI REG
- Backup the settings, before making temporary changes (and restore afterwards).
  Example: Contrib.SettingsExporter REG SOFTWARE\OpenRails-Saved\ORTS
```

Exporting from the Registry, when OpenRails.ini already exists (a backup is created):

```
C:\Users\roger\source\repos\rwf-openrails\Program>Contrib.SettingsExporter REG INI
Info: Loading from SOFTWARE\OpenRails\ORTS.
Info: Successfully loaded settings from HKEY_CURRENT_USER\SOFTWARE\OpenRails\ORTS.
Info: Backed up existing INI file as C:\Users\roger\source\repos\rwf-openrails\Program\OpenRails.ini.bak.
Info: Saving to C:\Users\roger\source\repos\rwf-openrails\Program\OpenRails.ini.
Info: Successfully saved to C:\Users\roger\source\repos\rwf-openrails\Program\OpenRails.ini.
```

**Note**: 
The actual output includes a list of UserSettings names. These are printed by the following code in `UserSettings.Save()`:
```
            foreach (var property in GetProperties())
                if (property.GetCustomAttributes(typeof(DoNotSaveAttribute), false).Length == 0)
                {
                    Console.WriteLine(property.Name, property.PropertyType);
                    Save(property.Name, property.PropertyType);
                }
```
I am not sure why that `Console.WriteLine` is there. The other settings classes don't have it.


Feature request: Missing (please add Elvas Tower, Trello, Launchpad link if one exists)

rwf-rr committed:
Changes needed for exporter.

rwf-rr committed:
Make path expansion consistent.

rwf-rr committed:
Add SettingsExporter (contrib).

rwf-rr committed:
Merge branch 'refactor-user-settings-location' into refactor-user-settings-loc/export-settings

rwf-rr committed:
Merge branch 'refactor-user-settings-location' into refactor-user-settings-loc/export-settings

rwf-rr committed:
Merge branch 'refactor-user-settings-location' into refactor-user-settings-loc/export-settings

rwf-rr committed:
Merge branch 'refactor-user-settings-location' into refactor-user-settings-loc/export-settings

rwf-rr committed:
Add more error handling and info.

rwf-rr committed:
Merge branch 'refactor-user-settings-location' into refactor-user-settings-loc/export-settings

rwf-rr committed:
Fix bug in refactoring of user settings.

Will be cherry-picked into refactor-user-settings-location.

rwf-rr committed:
Merge branch 'master' into refactor-user-settings-loc/export-settings

Included manual resolution of conflict in Source/ORTS.sln.

rwf-rr committed:
Fix minor formatting issues.

rwf-rr committed:
Update user manual for Settings Exporter.

SteelFill commented on Source/Documentation/Manual/appendices.rst:
```diff
@@ -738,12 +738,20 @@ INI File and User Settings
+and will only be used by that installation of Open Rails.
 
-Create an empty file OpenRails.ini in the same folder as OpenRails.exe and start Open Rails.
+Use the contributed "Settings Exporter" to copy the existing settings from the registry to the INI file.
+The Settings Exporter is accessible from the "Tools" button in the main window.
+The Settings Exporter may also be used to create a backup of the settings, or to copy the settings from
+the INI file back into the registry.
```
So, this feature is very useful and works as designed in my messing around, but I feel that you should advise (both here in the menu and in the help section of the command prompt) **closing all open instances of the OpenRails Menu** (OpenRails.exe) before running any settings export commands.

Whenever OpenRails.exe is closed, it _overwrites_ the settings data wherever it originally loaded settings from with the settings as configured in the options menu, and the options menu does not update in real time to reflect changes made to the settings data by other programs. So basically, OR will happily undo all changes made by the settings exporter if the settings exporter is run first, then OpenRails.exe is closed afterwards.

Technically it's fine to export from INI to REG if OpenRails.exe is running off the settings from the INI file, and it's fine to create a new INI file from REG while OpenRails.exe is running off the settings in the registry, but it's safer to just tell folks to close OpenRails.exe before hitting enter on any settings exporter command.

rwf-rr replied:
Thanks for the great suggestion. I will add the warning.

rwf-rr committed:
Add inffo that OpenRails must be closed for export.

Includes an unrelated change. The check for an INI file path was
inconsistent, using EndsWith for input and Contains for output.

SteelFill reviewed (approved):
All looks good, this will be a very handy tool going forward.
